samples/subgroup-scan.cpp

```diff
@@ -0,0 +1,208 @@
+  q.submit([&](sycl::handler& cgh) {
+    auto scans = in.template get_access<mode::read>(cgh);
+    auto elems = ends.template get_access<mode::discard_write>(cgh);
+
+    cgh.parallel_for<kernel_name<T, Op, class copy_ends>>(
+        sycl::range<1>(n_segments), [=](sycl::item<1> item) {
+          auto id = item.get_linear_id();
+          // Offset into the last element of each segment.
+          elems[item] = scans[(id + 1) * wgroup_size - 1];
+        });
+  });
+
+  // Recursively scan the array of last elements.
+  par_scan<T, Op>(ends, q);
+
+  // Add the results of the scan to each segment.
+  q.submit([&](sycl::handler& cgh) {
+    auto ends_scan = ends.template get_access<mode::read>(cgh);
+    auto data = in.template get_access<mode::read_write>(cgh);
+
+    cgh.parallel_for<kernel_name<T, Op, class add_ends>>(
+        // Work with one less work-group, since the first segment is correct.
+        sycl::nd_range<1>(input_size - wgroup_size, wgroup_size),
+        [=](sycl::nd_item<1> item) {
+          auto group = item.get_group_linear_id();
+          auto off_gid = item.get_global_linear_id() + wgroup_size;
+          data[off_gid] = Op{}(data[off_gid], ends_scan[group]);
+        });
+  });
+}
+
+/* Tests the scan with an addition operation, which is its most common use.
+ * Returns 0 if successful, a nonzero value otherwise. */
+int test_sum(sycl::queue& q) {
+  constexpr size_t size = 8192;
+
+  std::vector<int32_t> in(size);
+  std::iota(in.begin(), in.end(), 1);
+
+  std::vector<int32_t> sum(in.size());
+  {
+    sycl::buffer<int32_t, 1> buf(sycl::range<1>(in.size()));
+    buf.set_final_data(sum.data());
+    q.submit([&](sycl::handler& cgh) {
+      auto acc = buf.get_access<sycl::access::mode::write>(cgh);
+      cgh.copy(in.data(), acc);
+    });
+
+    par_scan<int32_t, sycl::experimental::plus<int32_t>>(buf, q);
+  }
+
+  std::vector<int32_t> test_sum(in.size());
+  std::partial_sum(in.begin(), in.end(), test_sum.begin());
+
+  auto equal = std::equal(sum.begin(), sum.end(), test_sum.begin());
+  if (!equal) {
+    std::cout << "SYCL sum computation incorrect! CPU Results:\n";
+    for (auto a : test_sum) {
+      std::cout << a << "\n";
+    }
+    std::cout << "\nSYCL results:\n";
+    for (auto a : sum) {
+      std::cout << a << "\n";
+    }
+    std::cout << std::endl;
+    return 1;
+  }
+
+  return 0;
+}
+
+int main() {
+  sycl::queue q{sycl::default_selector{}};
+  sycl::device d = q.get_device();
+
+  auto version = d.get_platform().get_info<sycl::info::platform::version>();
+  auto is_version = [&](std::string_view req_ver) {
+    return version.find(req_ver) != std::string::npos;
+  };
+  auto has_subgroups = d.is_host() || is_version("2.1") ||
+                       is_version("2.2") || is_version("3.0");
+  auto has_ext_subgroups = d.has_extension("cl_intel_subgroups") ||
+                           d.has_extension("cl_khr_subgroups");
+  if (!(has_subgroups || has_ext_subgroups)) {
```

Comment:
The 2 booleans are not needed anymore. I think we can just check that `d.get_info<cl::sycl::info::device::max_num_sub_groups>() > 0` now.

Reply:
This does not seem to work on the Intel GPU driver that I have on my laptop (though that is rather old, maybe I just need to update). Version 20.23.16988 for reference

Reply:
Ok, that's annoying. I may need to change that in SYCL-DNN.